FEATURE: Trashbin for the workspace module  / a.k.a. "EPIC: Trash"

pKallert opened the pull request:
## Trashbin for the workspace module 
This feature extends the current workspace module and adds a new view for each workspace where deleted Elements can be restored.
Deleted Elements in this case do not mean current changes in the workspace, but Elements that have been deleted in the past. Thanks to the new Content repository that is now possible. 



### New button in the workspace overview
Note: the button is only enabled if the workspace is up-to-date
<img width="1219" height="524" alt="Bildschirmfoto 2025-12-11 um 12 27 38" src="https://github.com/user-attachments/assets/09cd518c-872c-4e97-b4f5-193cf6b7b5f2" />

### The overview of deleted nodes in each workspace
<img width="1315" height="863" alt="Bildschirmfoto 2025-12-11 um 12 50 21" src="https://github.com/user-attachments/assets/8e2ee22c-9d41-43fe-bf39-80c261d255fa" />

### Searching for deleted nodes
<img width="1323" height="424" alt="Bildschirmfoto 2025-12-11 um 13 03 51" src="https://github.com/user-attachments/assets/a8b98b58-39b1-4240-a67b-e21a5136612e" />

### Pagination if there are more than 20 deleted elements 
<img width="1315" height="857" alt="Bildschirmfoto 2025-12-11 um 13 08 40" src="https://github.com/user-attachments/assets/2e5d9c5e-1b94-43af-967d-91e5933e1786" />

### Popup that shows which parent and child elements will additionally be restored 
<img width="1312" height="861" alt="Bildschirmfoto 2025-12-11 um 12 51 01" src="https://github.com/user-attachments/assets/be06de18-04a2-4a5f-ad11-f3a89b24537f" />
 
### Synchronize workspace button
If the workspace becomes outdated while the user is restoring elements, a button for syncing the workspace will be shown
<img width="1317" height="863" alt="Bildschirmfoto 2025-12-11 um 12 51 46" src="https://github.com/user-attachments/assets/1a4fb073-c7e9-4157-8ff3-f599dc80721a" />



## Upgrade instructions

This change introduces a new content repository projection. To enable the functionality when opening the trash bin the new tables need to be setup and also the projection subscription needs to be started and will already for a running Neos include all trash items if any are found in the events.

```
./flow cr:setup
./flow subscription:replay Neos.Workspace.Ui:TrashBinProjection
```

**Todo**
- [x] fix phpstan errors
- [x] fix most of the remaining todos and explain remaining todos well
- [x] add title labels to buttons in the ui


**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

pKallert committed:
Feature: Add basic restore module

pKallert committed:
Feature: Add module to composer file

pKallert committed:
Feature: Add basic table and frontend confirmation popups

pKallert committed:
Merge pull request #1 from pKallert/feature/add-frontend-modals-for-epic-trashtalk

Feature: Add basic table and frontend confirmation popups

mficzel commented:
A previous attempt to add deleted nodes to the node tree https://github.com/neos/neos-ui/pull/2974

pKallert committed:
Feature: List deleted nodes

pKallert committed:
Feature: Add Dimensionspacepoints to view

pKallert committed:
Feature: Add new table column for dimensions

Commit:
WIP: Implement trash bin

pKallert committed:
Merge branch 'epic-trashtalk' of https://github.com/pKallert/neos-development-collection into epic-trashtalk

Commit:
WIP: Implement TrashBinPopulationHook

pKallert committed:
Merge branch 'epic-trashtalk' of https://github.com/pKallert/neos-development-collection into epic-trashtalk

pKallert committed:
Feature: Move Restore UI files to workspace UI

pKallert committed:
Feature: make Trashbin work in new structure

pKallert committed:
Feature: Add pagination and Translations

pKallert committed:
Feature: Adjust pagination to use pagination object

pKallert committed:
Feature: Show basic list of restorable items

Commit:
Merge branch '9.0' into epic-trashtalk

Commit:
Merge branch '9.1' into epic-trashtalk

Commit:
Adjust test bootstrap

Commit:
WIP: Replace trash bin catchup hook with projection

Commit:
Add workspace support

Commit:
Add further restoration tests

Commit:
Reactivate restore module

Commit:
Introduce search terms in trash bin and add search tests

sstengel-vogel committed:
Add functionality Pagination

sstengel-vogel committed:
Add SearchTerm Field

sstengel-vogel committed:
change Pagination

sstengel-vogel committed:
change pagination behavior

sstengel-vogel committed:
change pagination behavior

sstengel-vogel committed:
change pagination

Commit:
only apply non-empty search terms

Commit:
keep SearchTerm in pagination

Commit:
Introduce sorting and keep page and search term where necessary

sstengel-vogel committed:
fix trash dialog call error

pKallert committed:
Feature: Make restore work

pKallert committed:
Feature: Add dropdown for chosing workspace

pKallert committed:
Feature: Make hard delete work

pKallert committed:
Feature: Add basic query for restoring parent elements

pKallert committed:
Feature: Show list of restorable elements

pKallert committed:
Merge branch '9.1' into epic-trashtalk

Sebobo commented on Neos.Workspace.Ui/Classes/Controller/RestoreController.php:
```diff
@@ -0,0 +1,405 @@
+        $contentRepository = $this->contentRepositoryRegistry->get($contentRepositoryId);
+
+        // @todo: resolve command(s) to result in removal of all soft-removed nodes on live (see GarbageCollector)
+        $nodeAggregate = $contentRepository->getContentGraph(WorkspaceName::forLive())->findNodeAggregateById($nodeAggregateId);
```
This is IMO a weird behaviour for the user:

1. I go to any workspace, e.g. "drafts"
2. I delete a node
3. I go to the trashbin
4. I delete forever
5. The node is deleted from live, even though I never published this change

pKallert replied:
Perhaps we can add the live workspace in the list workspace list. When showing the live workspace only the delete forever button is displayed as an option.
We could also go iteratively for now and only implement the restore function and think about the delete forever later on?

Sebobo replied:
Yes, focus on restore. Hard delete IMHO sounds more like a special pro feature and if I understood Bernhard correctly should only be for a small user group.

Sebobo committed:
BUGFIX: Forward workspace name to return updated entries after hard delete

Sebobo committed:
BUGFIX: Update dialogs after confirmation

Sebobo commented on Neos.Workspace.Ui/Classes/Controller/RestoreController.php:
```diff
@@ -0,0 +1,405 @@
+        //@todo: Check if parent restoration is needed
+        //@todo: the forwards do not work and do not show the expected values
+        if ($contentRepository->getContentGraph($workspaceName)->findParentNodeAggregates($nodeAggregateId)){
+            $this->forward(actionName: 'restoreNodeConfirmation', arguments: [
```
Can we not instead show all the information in the dialog instead of traversing up step by step? 
In many cases these could lead to 3-4 steps if a f.e. a multicolumn with an element was deleted.

pKallert replied:
Would you prefer to automatically also restore the parents? That would also restore the whole subtree of the parent and could lead to gigantic lists.

Sebobo replied:
The information about the parents is enough, we can then say that other content might also be restored which had been deleted with those parent nodes.

Sebobo commented on Neos.Workspace.Ui/Resources/Private/Fusion/Features/Workspace/Actions/Index.fusion:
```diff
@@ -1,8 +1,10 @@
 Neos.Workspace.Ui.WorkspaceController.index = Neos.Fusion:Component {
```
This one will never be rendered as the index redirects, right?

Sebobo commented on Neos.Workspace.Ui/Resources/Private/Fusion/Features/Restore/Components/NodesTableRow.fusion:
```diff
@@ -0,0 +1,96 @@
+          isDanger
+          title={private.i18n.id('restore.actions.hardDelete')}
+          icon="trash-alt icon-white"
+          disabled={!Security.hasAccess('Neos.Workspace.Ui:Backend.HardDeleteNodes')}
```
Hard delete should only be possible for nodes that exist and are soft deleted in the live workspace

pKallert replied:
Hard delete button is now only shown for live workspace. Since the live workspace is not shown in the overview, currently the hard delete functionality is never shown

Sebobo commented on Neos.Workspace.Ui/Resources/Private/Fusion/Features/Restore/Actions/Show.fusion:
```diff
@@ -0,0 +1,166 @@
+          </p>
+        </div>
+      </main>
+      <Neos.Workspace.Ui:Component.Footer>
```
The back button should be shown to allow going back to the review

pKallert replied:
Added a new button here

Sebobo commented:
After discussing the feature with @mhsdesign we have the following ideas to improve UX and make it possible to merge the functionality into 9.1 if we can get it done in time:

* Remove workspace dropdown, as it doesn't show the hierarchy from the main workspace list. If we keep the select box, only show the parent workspaces.
* Add the module footer with the back button to be able to return to the workspace list.
* All the buttons and actions need title attributes.
* Make the browser "back" button work to return to the workspace list.
* Adjust the dialog text and wordings to clearly differentiate the result of "restore element" from "discarding the deletion"
* Don't show "delete forever" for a workspace, this should be a separate view only for the "live" workspace, as "deleting forever" in a subworkspace is an action that is hard to understand as it can create unintended consequences from our point of view.
* Mid term (9.2 or 9.3): we might not even need the new view if the deleted elements in the changes list can be filtered and every deleted node has a "restore" button. 

Wdyt @nezaniel @pKallert ?

pKallert committed:
Feature: Remove Workspace switcher and custom menu item

pKallert committed:
Feature: Add back to overview button

pKallert committed:
Merge branch 'epic-trashtalk' of https://github.com/pKallert/neos-development-collection into epic-trashtalk

pKallert committed:
Feature: push History when changing to restore module, reorder search bar in restore view

pKallert committed:
Feature: Show hard delete only for live workspace

pKallert committed:
Feature: Remove Sync functionality from php part

pKallert commented:
I addressed most of the comments apart form this one: 
Adjust the dialog text and wordings to clearly differentiate the result of "restore element" from "discarding the deletion" 

Not sure what is meant with the differentiation here?

Also to think about: Should we show a banner to tell the user to sync the workspace if the workspace is outdated? Could be that there are more soft deletions in the base workspace that can be synced in.

Sebobo commented:
> I addressed most of the comments apart form this one: Adjust the dialog text and wordings to clearly differentiate the result of "restore element" from "discarding the deletion"
> 
> Not sure what is meant with the differentiation here?
> 
> Also to think about: Should we show a banner to tell the user to sync the workspace if the workspace is outdated? Could be that there are more soft deletions in the base workspace that can be synced in.

A sync shouldn't lead to more nodes to be restorable IMO. They are part of the base workspace and should be restored there.

The difference between restore and discard is relevant from what I understood, as one ONLY undoes the deletion, while a discard change, would discard all changes to the node even property changes.

pKallert committed:
Fix: tests

pKallert committed:
Feature: Show Ancestors in restore confirmation dialogue

pKallert committed:
Feature: Remove hardDelete function from Restore UI

pKallert committed:
Fix: Remove unused WorkspaceOutdated function

pKallert committed:
Feature: Remove unused param

mhsdesign commented:
We discussed this change in today's developer meeting how this feature fits into the next release.

Due to limited time we want to simplify the functionality for now by excluding the hard delete function from the live workspace.
The idea is valid but its a dangerous capability we add here - even if only for admins. The reason we think we need this function at some point is that the garbage collection can easily blocked "forever" by a stale editor workspace - but the actual problem would be that: A stale editor workspace is already bad in two ways: Old node (regular) rows are not cleaned up in the database and relogin (and rebase) will take long.

But as the hard deletion is a feature of the CR and we would have strategies in the Neos Ui to handle errors on rebase via a forced rebase. We might as well offer this as a functionality in the ui. In case the garbage collection does not work which should not happen (tm)). Though a backend module offering such hard delete should as we discussed list all possibly "impending conflicts" so the admin understands which workspaces he might turn unusable in the worst case.

Thank you paula for removing that capability for now then. I guess we could already also open a WIP Pr or revert https://github.com/neos/neos-development-collection/pull/5542/commits/1b20dbf2c0d0f69b7e83de6e2fb31ab81168fa0a at any point we like to reintroduce this feature and polish it.

mhsdesign committed:
PATCH: Remove obsolete `Neos.Workspace.Ui:Backend.HardDeleteNodes` privilege

mhsdesign committed:
PATCH: Simplify `Neos.Workspace.Ui` `FeatureContext` after copying from Neos.Neos

mhsdesign committed:
PATCH: Unify whitespaces in fusion files (manual linting *g)

mhsdesign committed:
PATCH: Remove obsolete method `findByContentStreamIdAndChangeType` and `ChangeType` enum

mhsdesign committed:
PATCH: Rename `Neos.Workspace.Ui:Backend.Access` to be more precise to `RestoreNodes`

mhsdesign committed:
PATCH: Optimise imports in new restore Ui

mhsdesign committed:
PATCH: Adjust `@internal` notice and add to all trash bing things

mhsdesign committed:
PATCH: Fix type annotation mismatch in the `TrashBinProjection`

mhsdesign committed:
PATCH: Apply phpcs to new Neos.Workspace.Ui restore and trash bin

mhsdesign commented:
Okay i tested the functionality a little and did a simple first review on how the code looks like. It seems like phpstan complains a lot now with still 19 errors and also there are many (9) todos sorely in the new restore controller.

We dont have proper linting for fusion but there where some odd whitespaces so i fixed them by hand. I also removed a little more obsolete code in https://github.com/neos/neos-development-collection/pull/5542/commits/1b70bd15f4ff58ab4ec2f387f332b54b8072d82b, https://github.com/neos/neos-development-collection/pull/5542/commits/709c60e608d018f6e995dd2ffa25a687c219f935 and https://github.com/neos/neos-development-collection/pull/5542/commits/d1ff36a7b0ad9a4a8dda15f192ac024492fe8fb0 as well as did some more costmetic little changes here and there like adjusting comments.

For php i enabled php-cs and let it fix the code style https://github.com/neos/neos-development-collection/pull/5542/commits/19fa933db1d09241833f6398b371fa5ce5e944d1 as well as optimised the imports a little with phpstorm https://github.com/neos/neos-development-collection/pull/5542/commits/0776af6d8922efbd29eca8f9875c88dc0ced6066.

EDIT: With https://github.com/neos/neos-development-collection/pull/5542/commits/3641a71e434b68860df7c4bb77e6b316dc21f84c and before i fixed a few phpstan errors so its now only 13 sorely in the `RestoreController` which already needs some love i guess looking at its todo comments ...

do you need also help to bring the `RestoreController` to full glory then? @nezaniel @pKallert

mhsdesign committed:
PATCH: Fix phpstan in view models

mhsdesign committed:
PATCH: Handle possible `null` case in event metadata for user-id and delete-time

further the `TrashItem` uses the CR UserId value object as we cannot be sure that its a Neos user id. For example the `system` CR user is not and this distinction makes this case more explicit.

Commit:
5542 - translate node type names

Commit:
5542 - properly resolve messages, labels, ancestors and descendants

pKallert committed:
Fix: Wrong table break for variants

pKallert committed:
Fix: labels, center icons, back button url format

pKallert committed:
Feature: Adjust styling of confirmation window list

pKallert committed:
Feature: disable restore button in list for outdated workspaces

pKallert committed:
Feature: Add sync workspace butotn

pKallert committed:
Fix: Linter

pKallert committed:
Feature: Translate Workspace label and remove unused parentRestore

pKallert committed:
Feature: Show Search field and message if there are no results for the search

pKallert committed:
Fix: Linter

Commit:
Pacify linter

Commit:
Also restore all removed ancestors

Commit:
Prevent duplicate ancestor restoration commands

Sebobo committed:
BUGFIX: Prevent htmx from duplicating changes list on restore failure

Sebobo reviewed (approved):
Works on my machine <3

mhsdesign reviewed (approved):
Perfekt. Thanks for all the adjustments the code looks good and it works as far as i tested it ;) (Also thanks for the tests <3)

Sebobo committed:
TASK: Remove unnecessary forward after throwing status
